---
output:
  md_document:
    variant: markdown_github
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r, echo = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "README-"
)
```

# MTGOsc

## Extracting gene modules from single cell RNA-seq cell clusters

MTGO-SC is an adaptation for scRNA-seq (scRNA-seq) of [MTGO](https://gitlab.com/d1vella/MTGO), a biological network module detection algorithm. MTGO-SC integrates external gene annotations, such as the Gene Ontology terms or Reactome pathways with the gene expression networks obtained from single-cell DGE matrices.

## A post-processing step after cell clustering

The typical scRNA-seq pipeline is designed to group cells into meaningful clusters, representing cells of similar (sub)type or stare. MTGO-SC extracts the gene interaction network from each cluster, and exctract gene functional modules labeled with an annotation from the source provided by the user (for example, Reactome pathways).

## Example of MTGO-SC application
A practical example of application on [MCA](http://bis.zju.edu.cn/MCA/) cluster, along with enrichment and term literature search, is provided in the Vignette.
