<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1">

<meta name="author" content="Nelson Nazzicari, Simone Marini, Danila Vella" />

<meta name="date" content="2019-07-22" />

<title>MTGOsc: MTGO analysis on single cells</title>



<style type="text/css">code{white-space: pre;}</style>
<style type="text/css" data-origin="pandoc">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
code > span.kw { color: #007020; font-weight: bold; } /* Keyword */
code > span.dt { color: #902000; } /* DataType */
code > span.dv { color: #40a070; } /* DecVal */
code > span.bn { color: #40a070; } /* BaseN */
code > span.fl { color: #40a070; } /* Float */
code > span.ch { color: #4070a0; } /* Char */
code > span.st { color: #4070a0; } /* String */
code > span.co { color: #60a0b0; font-style: italic; } /* Comment */
code > span.ot { color: #007020; } /* Other */
code > span.al { color: #ff0000; font-weight: bold; } /* Alert */
code > span.fu { color: #06287e; } /* Function */
code > span.er { color: #ff0000; font-weight: bold; } /* Error */
code > span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
code > span.cn { color: #880000; } /* Constant */
code > span.sc { color: #4070a0; } /* SpecialChar */
code > span.vs { color: #4070a0; } /* VerbatimString */
code > span.ss { color: #bb6688; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { color: #19177c; } /* Variable */
code > span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code > span.op { color: #666666; } /* Operator */
code > span.bu { } /* BuiltIn */
code > span.ex { } /* Extension */
code > span.pp { color: #bc7a00; } /* Preprocessor */
code > span.at { color: #7d9029; } /* Attribute */
code > span.do { color: #ba2121; font-style: italic; } /* Documentation */
code > span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code > span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code > span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */

</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    for (var j = 0; j < rules.length; j++) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") continue;
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') continue;
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>



<link href="data:text/css;charset=utf-8,body%20%7B%0Abackground%2Dcolor%3A%20%23fff%3B%0Amargin%3A%201em%20auto%3B%0Amax%2Dwidth%3A%20700px%3B%0Aoverflow%3A%20visible%3B%0Apadding%2Dleft%3A%202em%3B%0Apadding%2Dright%3A%202em%3B%0Afont%2Dfamily%3A%20%22Open%20Sans%22%2C%20%22Helvetica%20Neue%22%2C%20Helvetica%2C%20Arial%2C%20sans%2Dserif%3B%0Afont%2Dsize%3A%2014px%3B%0Aline%2Dheight%3A%201%2E35%3B%0A%7D%0A%23header%20%7B%0Atext%2Dalign%3A%20center%3B%0A%7D%0A%23TOC%20%7B%0Aclear%3A%20both%3B%0Amargin%3A%200%200%2010px%2010px%3B%0Apadding%3A%204px%3B%0Awidth%3A%20400px%3B%0Aborder%3A%201px%20solid%20%23CCCCCC%3B%0Aborder%2Dradius%3A%205px%3B%0Abackground%2Dcolor%3A%20%23f6f6f6%3B%0Afont%2Dsize%3A%2013px%3B%0Aline%2Dheight%3A%201%2E3%3B%0A%7D%0A%23TOC%20%2Etoctitle%20%7B%0Afont%2Dweight%3A%20bold%3B%0Afont%2Dsize%3A%2015px%3B%0Amargin%2Dleft%3A%205px%3B%0A%7D%0A%23TOC%20ul%20%7B%0Apadding%2Dleft%3A%2040px%3B%0Amargin%2Dleft%3A%20%2D1%2E5em%3B%0Amargin%2Dtop%3A%205px%3B%0Amargin%2Dbottom%3A%205px%3B%0A%7D%0A%23TOC%20ul%20ul%20%7B%0Amargin%2Dleft%3A%20%2D2em%3B%0A%7D%0A%23TOC%20li%20%7B%0Aline%2Dheight%3A%2016px%3B%0A%7D%0Atable%20%7B%0Amargin%3A%201em%20auto%3B%0Aborder%2Dwidth%3A%201px%3B%0Aborder%2Dcolor%3A%20%23DDDDDD%3B%0Aborder%2Dstyle%3A%20outset%3B%0Aborder%2Dcollapse%3A%20collapse%3B%0A%7D%0Atable%20th%20%7B%0Aborder%2Dwidth%3A%202px%3B%0Apadding%3A%205px%3B%0Aborder%2Dstyle%3A%20inset%3B%0A%7D%0Atable%20td%20%7B%0Aborder%2Dwidth%3A%201px%3B%0Aborder%2Dstyle%3A%20inset%3B%0Aline%2Dheight%3A%2018px%3B%0Apadding%3A%205px%205px%3B%0A%7D%0Atable%2C%20table%20th%2C%20table%20td%20%7B%0Aborder%2Dleft%2Dstyle%3A%20none%3B%0Aborder%2Dright%2Dstyle%3A%20none%3B%0A%7D%0Atable%20thead%2C%20table%20tr%2Eeven%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0A%7D%0Ap%20%7B%0Amargin%3A%200%2E5em%200%3B%0A%7D%0Ablockquote%20%7B%0Abackground%2Dcolor%3A%20%23f6f6f6%3B%0Apadding%3A%200%2E25em%200%2E75em%3B%0A%7D%0Ahr%20%7B%0Aborder%2Dstyle%3A%20solid%3B%0Aborder%3A%20none%3B%0Aborder%2Dtop%3A%201px%20solid%20%23777%3B%0Amargin%3A%2028px%200%3B%0A%7D%0Adl%20%7B%0Amargin%2Dleft%3A%200%3B%0A%7D%0Adl%20dd%20%7B%0Amargin%2Dbottom%3A%2013px%3B%0Amargin%2Dleft%3A%2013px%3B%0A%7D%0Adl%20dt%20%7B%0Afont%2Dweight%3A%20bold%3B%0A%7D%0Aul%20%7B%0Amargin%2Dtop%3A%200%3B%0A%7D%0Aul%20li%20%7B%0Alist%2Dstyle%3A%20circle%20outside%3B%0A%7D%0Aul%20ul%20%7B%0Amargin%2Dbottom%3A%200%3B%0A%7D%0Apre%2C%20code%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0Aborder%2Dradius%3A%203px%3B%0Acolor%3A%20%23333%3B%0Awhite%2Dspace%3A%20pre%2Dwrap%3B%20%0A%7D%0Apre%20%7B%0Aborder%2Dradius%3A%203px%3B%0Amargin%3A%205px%200px%2010px%200px%3B%0Apadding%3A%2010px%3B%0A%7D%0Apre%3Anot%28%5Bclass%5D%29%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0A%7D%0Acode%20%7B%0Afont%2Dfamily%3A%20Consolas%2C%20Monaco%2C%20%27Courier%20New%27%2C%20monospace%3B%0Afont%2Dsize%3A%2085%25%3B%0A%7D%0Ap%20%3E%20code%2C%20li%20%3E%20code%20%7B%0Apadding%3A%202px%200px%3B%0A%7D%0Adiv%2Efigure%20%7B%0Atext%2Dalign%3A%20center%3B%0A%7D%0Aimg%20%7B%0Abackground%2Dcolor%3A%20%23FFFFFF%3B%0Apadding%3A%202px%3B%0Aborder%3A%201px%20solid%20%23DDDDDD%3B%0Aborder%2Dradius%3A%203px%3B%0Aborder%3A%201px%20solid%20%23CCCCCC%3B%0Amargin%3A%200%205px%3B%0A%7D%0Ah1%20%7B%0Amargin%2Dtop%3A%200%3B%0Afont%2Dsize%3A%2035px%3B%0Aline%2Dheight%3A%2040px%3B%0A%7D%0Ah2%20%7B%0Aborder%2Dbottom%3A%204px%20solid%20%23f7f7f7%3B%0Apadding%2Dtop%3A%2010px%3B%0Apadding%2Dbottom%3A%202px%3B%0Afont%2Dsize%3A%20145%25%3B%0A%7D%0Ah3%20%7B%0Aborder%2Dbottom%3A%202px%20solid%20%23f7f7f7%3B%0Apadding%2Dtop%3A%2010px%3B%0Afont%2Dsize%3A%20120%25%3B%0A%7D%0Ah4%20%7B%0Aborder%2Dbottom%3A%201px%20solid%20%23f7f7f7%3B%0Amargin%2Dleft%3A%208px%3B%0Afont%2Dsize%3A%20105%25%3B%0A%7D%0Ah5%2C%20h6%20%7B%0Aborder%2Dbottom%3A%201px%20solid%20%23ccc%3B%0Afont%2Dsize%3A%20105%25%3B%0A%7D%0Aa%20%7B%0Acolor%3A%20%230033dd%3B%0Atext%2Ddecoration%3A%20none%3B%0A%7D%0Aa%3Ahover%20%7B%0Acolor%3A%20%236666ff%3B%20%7D%0Aa%3Avisited%20%7B%0Acolor%3A%20%23800080%3B%20%7D%0Aa%3Avisited%3Ahover%20%7B%0Acolor%3A%20%23BB00BB%3B%20%7D%0Aa%5Bhref%5E%3D%22http%3A%22%5D%20%7B%0Atext%2Ddecoration%3A%20underline%3B%20%7D%0Aa%5Bhref%5E%3D%22https%3A%22%5D%20%7B%0Atext%2Ddecoration%3A%20underline%3B%20%7D%0A%0Acode%20%3E%20span%2Ekw%20%7B%20color%3A%20%23555%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%0Acode%20%3E%20span%2Edt%20%7B%20color%3A%20%23902000%3B%20%7D%20%0Acode%20%3E%20span%2Edv%20%7B%20color%3A%20%2340a070%3B%20%7D%20%0Acode%20%3E%20span%2Ebn%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Efl%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Ech%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Est%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Eco%20%7B%20color%3A%20%23888888%3B%20font%2Dstyle%3A%20italic%3B%20%7D%20%0Acode%20%3E%20span%2Eot%20%7B%20color%3A%20%23007020%3B%20%7D%20%0Acode%20%3E%20span%2Eal%20%7B%20color%3A%20%23ff0000%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%0Acode%20%3E%20span%2Efu%20%7B%20color%3A%20%23900%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%20code%20%3E%20span%2Eer%20%7B%20color%3A%20%23a61717%3B%20background%2Dcolor%3A%20%23e3d2d2%3B%20%7D%20%0A" rel="stylesheet" type="text/css" />




</head>

<body>




<h1 class="title toc-ignore">MTGOsc: MTGO analysis on single cells</h1>
<h4 class="author">Nelson Nazzicari, Simone Marini, Danila Vella</h4>
<h4 class="date">2019-07-22</h4>


<div id="TOC">
<ul>
<li><a href="#example-data-bladder-basal-epithelial-cells">Example data: bladder basal epithelial cells</a></li>
<li><a href="#calling-mtgosc">Calling MTGOsc</a></li>
<li><a href="#gene-module-enrichment-on-reactome">Gene module enrichment on Reactome</a></li>
<li><a href="#comparing-network-extraction-methods">Comparing network extraction methods</a></li>
</ul>
</div>

<!-- 
PLEASE NOTE

This is a vignette and it's not rebuilt automatically with the package.
To do so please use the following command:

devtools::build_vignettes()

And *then* rebuild the package, so it updates the doc.
-->
<div id="example-data-bladder-basal-epithelial-cells" class="section level2">
<h2>Example data: bladder basal epithelial cells</h2>
<p>MTGOsc comes with some example data from the Mouse Atlas in the form of a (simplified for storage reasons) Seurat object. Let’s start with loading Seurat and MTGO:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(Seurat)
<span class="kw">library</span>(MTGOsc)</code></pre></div>
<p>Loading MTGOsc also loads three objects: <em>bladder</em>, <em>markers</em>, and <em>mouse.pathways</em>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co">#this is a (simplified) Seurat object</span>
<span class="kw">print</span>(bladder)</code></pre></div>
<pre><code>## An old seurat object
##  269 genes across 1028 samples</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co">#this come from applying Seurat function FindAllMarkers() to the bladder dataset</span>
<span class="kw">head</span>(markers)</code></pre></div>
<pre><code>##                p_val avg_logFC pct.1 pct.2     p_val_adj
## krt15  1.523860e-202  1.959457 0.994 0.269 2.070317e-198
## igfbp2 3.880320e-196  1.857702 1.000 0.322 5.271803e-192
## trf    2.219344e-186  1.747788 0.896 0.189 3.015200e-182
## krt5   1.329655e-180  1.730088 0.976 0.276 1.806470e-176
## gsdmc2 1.235398e-172  1.711701 0.908 0.208 1.678412e-168
## gsto1  2.506894e-172  1.629429 0.985 0.299 3.405866e-168
##                               cluster   gene
## krt15  Basal epithelial cell(Bladder)  krt15
## igfbp2 Basal epithelial cell(Bladder) igfbp2
## trf    Basal epithelial cell(Bladder)    trf
## krt5   Basal epithelial cell(Bladder)   krt5
## gsdmc2 Basal epithelial cell(Bladder) gsdmc2
## gsto1  Basal epithelial cell(Bladder)  gsto1</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co">#this is a simple gene -&gt; pathway dictionary</span>
<span class="kw">head</span>(mouse.pathways)</code></pre></div>
<pre><code>##    gene                                  pathway
## 1 43160                               Metabolism
## 2 43160                    Biological oxidations
## 3 43160 Phase I - Functionalization of compounds
## 4 43161                               Metabolism
## 5 43161                    Biological oxidations
## 6 43161 Phase I - Functionalization of compounds</code></pre>
<p>Bladder object contains three clusters:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">table</span>(bladder<span class="op">@</span>ident)</code></pre></div>
<pre><code>## 
## Basal epithelial cell(Bladder) Stromal cell_Dpt high(Bladder) 
##                            327                            651 
##         Umbrella cell(Bladder) 
##                             50</code></pre>
</div>
<div id="calling-mtgosc" class="section level2">
<h2>Calling MTGOsc</h2>
<p>For this tutorial we’ll focus on Basal epitelial cells cluster, which is of intermediate size. We create two “selector” arrays, one for cells and one for genes.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">cells.selected =<span class="st"> </span>bladder<span class="op">@</span>ident <span class="op">==</span><span class="st"> 'Basal epithelial cell(Bladder)'</span>
genes.selected =<span class="st"> </span><span class="kw">subset</span>(markers, cluster <span class="op">==</span><span class="st"> 'Basal epithelial cell(Bladder)'</span>)<span class="op">$</span>gene

my.bladder =<span class="st"> </span>bladder
my.bladder<span class="op">@</span>data =<span class="st"> </span>my.bladder<span class="op">@</span>data[genes.selected, cells.selected]</code></pre></div>
<p>MTGOsc needs a folder where to save temporary files and final results.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">root =<span class="st"> </span><span class="kw">tempdir</span>() <span class="co">#change this to your preferred local path</span>
<span class="kw">dir.create</span>(root, <span class="dt">recursive =</span> <span class="ot">TRUE</span>, <span class="dt">showWarnings =</span> <span class="ot">FALSE</span>)</code></pre></div>
<p>We are now ready to start with MTGOsc:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co">#building a genes-pathways dictionary</span>
dict =<span class="st"> </span><span class="kw">write.dictionary</span>(<span class="dt">genes=</span>mouse.pathways<span class="op">$</span>gene, 
                       <span class="dt">terms =</span> mouse.pathways<span class="op">$</span>pathway, <span class="dt">outfolder =</span> root)

<span class="co">#computign gene coexpression (default function is 'cor')</span>
coexp =<span class="st"> </span><span class="kw">write.coexpressionMatrix</span>(<span class="dt">geneExpression =</span> my.bladder<span class="op">@</span>data, <span class="dt">outfolder =</span> root)

<span class="co">#thinning coexpression network via scale free criterion</span>
edges =<span class="st"> </span><span class="kw">write.edges</span>(<span class="dt">coexpression =</span> coexp, <span class="dt">outfolder =</span> root, 
                   <span class="dt">keep.weights =</span> <span class="ot">FALSE</span>, <span class="dt">fun =</span> thinning_scale_free)</code></pre></div>
<pre><code>## gamma: 2.98456061941471 (target:2)
## threshold: 0.9
## network edges: 18</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co">#writing a parameter file, useful for MGTO</span>
<span class="kw">write.paramFile</span>(<span class="dt">outfolder =</span> root)</code></pre></div>
<p>All the groundwork is done and we can invoke MTGO:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co">#actual call to MTGO</span>
<span class="co">#call.MTGO(outfolder = root, verbose = TRUE) #this prints several log messages</span>
 
<span class="co">#building and saving representation of resulting network</span>
<span class="co">#network.collapsed = export.network.modules(infolder = root, collapse.modules = TRUE) </span>
<span class="co">#network.full = export.network.modules(infolder = root, collapse.modules = FALSE) </span></code></pre></div>
<p>At this point networks are saved on disk, in “Network” subfolder, in the form of two html files. In the first one (ClusterNetwork.html) each functional module is collapsed to a single node, the second one (FullNetwork.html) where each gene is represented and functional modules are color coded.</p>
</div>
<div id="gene-module-enrichment-on-reactome" class="section level2">
<h2>Gene module enrichment on Reactome</h2>
<p>Here we look for Reactome pathway enrichment of the genes constituting the thinned network. This procedure is complementary to the exctraction of Reactome pathways by MTGO-SC.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># load libraries for gene enrichment on Reactome (those are on Bioconductor, not CRAN)</span>
<span class="kw">library</span>(ReactomePA)      
<span class="kw">library</span>(clusterProfiler)
<span class="kw">library</span>(org.Mm.eg.db)

<span class="co">#a support function to take care of gene upper/lower case convention</span>
firstup =<span class="st"> </span><span class="cf">function</span>(x) {
  x =<span class="st"> </span><span class="kw">tolower</span>(x)
  <span class="kw">substr</span>(x, <span class="dv">1</span>, <span class="dv">1</span>) =<span class="st"> </span><span class="kw">toupper</span>(<span class="kw">substr</span>(x, <span class="dv">1</span>, <span class="dv">1</span>))
  <span class="kw">return</span>(x)
}

<span class="co">#the list of all genes involved in the cluster</span>
genes =<span class="st"> </span><span class="kw">unique</span>(<span class="kw">c</span>(<span class="kw">as.character</span>(edges<span class="op">$</span>gene1), <span class="kw">as.character</span>(edges<span class="op">$</span>gene2)))

<span class="co">#correct casing of gene names</span>
genes =<span class="st"> </span><span class="kw">firstup</span>(genes)

<span class="co">#translating gene names to ENTREZID via org.Mm.eg.db database</span>
genes =<span class="st"> </span><span class="kw">bitr</span>(genes, <span class="dt">fromType=</span><span class="st">&quot;SYMBOL&quot;</span>, <span class="dt">toType=</span><span class="st">&quot;ENTREZID&quot;</span>, <span class="dt">OrgDb=</span><span class="st">&quot;org.Mm.eg.db&quot;</span>)

<span class="co">#the actual enrichment</span>
enriched =<span class="st"> </span><span class="kw">enrichPathway</span>(<span class="dt">gene=</span>genes<span class="op">$</span>ENTREZID, <span class="dt">pvalueCutoff=</span><span class="fl">0.05</span>, 
                         <span class="dt">readable=</span>T, <span class="dt">organism =</span> <span class="st">&quot;mouse&quot;</span>, <span class="dt">pAdjustMethod =</span> <span class="st">&quot;BH&quot;</span>)</code></pre></div>
<!-- ## Enriched term literature search with RISmed -->
</div>
<div id="comparing-network-extraction-methods" class="section level2">
<h2>Comparing network extraction methods</h2>
<p>In the previous example we extracted the gene interaction network that is fed to MTGO via a two step approach:</p>
<ol style="list-style-type: decimal">
<li>computing gene coexpression (default function: <em>cor</em>)</li>
<li>network thinning (we used the function <em>scale_free_threshold</em>)</li>
</ol>
<p>MTGOsc package offers several options for both steps, and it is always possible for the user to write and use their own functions. As such we propose an objective approach based on an (external) ground truth to select the best combination of functions.</p>
<p>In general terms, we’ll select the method that maximise the <em>Affinity Score</em> (AS), which is a metric of how close the extracted network is to a selected external ground truth. As such, the first thing we’ll need is said ground truth, which luckily is already saved in MTGOsc:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co">#gGT is an iGraph object derived from four different sources:</span>
<span class="co"># - the CORUM database [1]</span>
<span class="co"># - a protein interaction map [2]</span>
<span class="co"># - String and Reactome databases [3]  </span>
<span class="kw">summary</span>(gGT)</code></pre></div>
<pre><code>## IGRAPH 2229188 UN-- 14324 1007972 -- 
## + attr: name (v/c)</code></pre>
<p>For more details see the following references:</p>
<p>1 Giurgiu M, Reinhard J, Brauner B, Dunger-Kaltenbach I, Fobo G, Frishman G, Montrone C, Ruepp A. CORUM: the comprehensive resource of mammalian protein complexes-2019. Nucleic Acids Res. 2018 Oct 24. doi: 10.1093/nar/gky973. [Epub ahead of print] 30357367</p>
<p>2 Ramani, Arun K., et al. “A map of human protein interactions derived from co‐expression of human mRNAs and their orthologs.” Molecular systems biology 4.1 (2008): 180.</p>
<p>3 Skinnider, Michael A., Jordan W. Squair, and Leonard J. Foster. “Evaluating measures of association for single-cell transcriptomics.” Nature methods 16.5 (2019): 381.</p>
<p>For the sake of this tutorial we’ll compare two methods for extracting the interaction network:</p>
<ul>
<li>METHOD 1:
<ul>
<li>Gene coexpression computed via Pearson’s correlation (default <em>cor</em> function)</li>
<li>Network thinning to maximise scale free fit (<em>thinning_scale_free</em> function)</li>
</ul></li>
<li>METHOD 2:
<ul>
<li>Gene coexpression computed via proportionality metrics phi (<em>coexpr_propr</em> function)</li>
<li>Network thinning to keep the top ten percent of all gene-gene interactions (<em>thinning_percentile</em> function)</li>
</ul></li>
</ul>
<p>Extracted networks must be transformed in <em>iGraph</em> object, starting from the edges object, as follow:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co">#Edges for Method 1, computed above</span>
edges.M1 =<span class="st"> </span>edges

<span class="co">#Edges for Method 2, on the same data</span>
coexp.M2 =<span class="st"> </span><span class="kw">write.coexpressionMatrix</span>(
  <span class="dt">geneExpression =</span> my.bladder<span class="op">@</span>data, <span class="dt">outfolder =</span> root, <span class="dt">fun =</span> coexpr_propr, <span class="dt">metric =</span> <span class="st">'phs'</span>, <span class="dt">symmetrize =</span> <span class="ot">TRUE</span>, <span class="dt">overwrite =</span> <span class="ot">TRUE</span>)
edges.M2 =<span class="st"> </span><span class="kw">write.edges</span>(<span class="dt">coexpression =</span> coexp.M2, <span class="dt">outfolder =</span> root, 
                   <span class="dt">keep.weights =</span> <span class="ot">FALSE</span>, <span class="dt">fun =</span> thinning_percentile, <span class="dt">overwrite =</span> <span class="ot">TRUE</span>)

<span class="co">#transforming into iGraph</span>
network.M1 =<span class="st"> </span>igraph<span class="op">::</span><span class="kw">graph_from_edgelist</span>(<span class="kw">as.matrix</span>(edges.M1),<span class="dt">directed =</span> <span class="ot">FALSE</span>)
network.M2 =<span class="st"> </span>igraph<span class="op">::</span><span class="kw">graph_from_edgelist</span>(<span class="kw">as.matrix</span>(edges.M2),<span class="dt">directed =</span> <span class="ot">FALSE</span>)</code></pre></div>
<p>The Affinity Scores can be directly computed as follows:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co">#support function to compute Affinity Score between two networks</span>
compute_AS =<span class="st"> </span><span class="cf">function</span>(N1, N2){
  E.N1 =<span class="st"> </span><span class="kw">length</span>(igraph<span class="op">::</span><span class="kw">E</span>(N1))
  E.N2 =<span class="st"> </span><span class="kw">length</span>(igraph<span class="op">::</span><span class="kw">E</span>(N2))
  overlap =<span class="st"> </span><span class="kw">length</span>(igraph<span class="op">::</span><span class="kw">E</span>(igraph<span class="op">::</span><span class="kw">intersection</span>(N1, N2)))
  
  <span class="co">#the formula is [overlap^2] / (edges1 * edges2)</span>
  <span class="co">#but for better computation we implement it as</span>
  res =<span class="st"> </span>(overlap <span class="op">/</span><span class="st"> </span>E.N1) <span class="op">*</span><span class="st"> </span>(overlap <span class="op">/</span><span class="st"> </span>E.N2)
  
  <span class="co">#and we are done</span>
  <span class="kw">return</span>(res)
}

<span class="co">#ready to compute Affinity Scores for the examples</span>
AS.M1 =<span class="st"> </span><span class="kw">compute_AS</span>(network.M1, gGT)
AS.M2 =<span class="st"> </span><span class="kw">compute_AS</span>(network.M2, gGT)</code></pre></div>
<p>However they cannot be directly compared yet. For a proper comparison we should compute for each extracted network a random population of scrambled networks and then compare each AS to the distribution of scores. This will allow us to compute Z-scores and consequently P values associated to each extracted method. We can then (finally!) compare the logarithms of the P-values.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co">#numerosity for each random population</span>
N =<span class="st"> </span><span class="dv">100</span> 

<span class="co">#room for Affinity Scores for the two scrambled networks populations</span>
AS.distr.M1 =<span class="st"> </span><span class="kw">c</span>()
AS.distr.M2 =<span class="st"> </span><span class="kw">c</span>()

<span class="co">#doing the computation</span>
<span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span><span class="dv">100</span>){
  <span class="co">#creating a random scramble of both networks</span>
  random.network.M1 =<span class="st"> </span><span class="kw">rewire</span>(network.M1,<span class="dt">with=</span><span class="kw">keeping_degseq</span>(<span class="dt">niter =</span> <span class="kw">vcount</span>(network.M1) <span class="op">*</span><span class="st"> </span><span class="dv">10</span>,<span class="dt">loops =</span> <span class="ot">FALSE</span>))
  random.network.M2 =<span class="st"> </span><span class="kw">rewire</span>(network.M2,<span class="dt">with=</span><span class="kw">keeping_degseq</span>(<span class="dt">niter =</span> <span class="kw">vcount</span>(network.M2) <span class="op">*</span><span class="st"> </span><span class="dv">10</span>,<span class="dt">loops =</span> <span class="ot">FALSE</span>))
  
  <span class="co">#computing the new Affinity Scores</span>
  AS.distr.M1[i] =<span class="st"> </span><span class="kw">compute_AS</span>(random.network.M1, gGT)
  AS.distr.M2[i] =<span class="st"> </span><span class="kw">compute_AS</span>(random.network.M2, gGT)
}

<span class="co">#Computing Z scores</span>
Z.M1 =<span class="st"> </span>(AS.M1 <span class="op">-</span><span class="st"> </span><span class="kw">mean</span>(AS.distr.M1)) <span class="op">/</span><span class="st"> </span><span class="kw">sd</span>(AS.distr.M1)
Z.M2 =<span class="st"> </span>(AS.M2 <span class="op">-</span><span class="st"> </span><span class="kw">mean</span>(AS.distr.M2)) <span class="op">/</span><span class="st"> </span><span class="kw">sd</span>(AS.distr.M2)

<span class="co">#Computing -log(P values)</span>
logP.M1 =<span class="st"> </span><span class="op">-</span><span class="kw">log</span>(<span class="kw">pnorm</span>(<span class="op">-</span>Z.M1))
logP.M2 =<span class="st"> </span><span class="op">-</span><span class="kw">log</span>(<span class="kw">pnorm</span>(<span class="op">-</span>Z.M2))</code></pre></div>
<p>At this point the methods can be compared directly. In our case Method 2 shows higher values and should thus be preferred.</p>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
